### PR TITLE
Store browser accept language header on users

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This feature allows the system to store relevant visit info for signed in users, relating to ahoy visits, but in an efficient manner that relates visits to unique contexts.

This helps us collect browser language preferences to ultimately help us make better default choices about user language preferences, as a first step.

## Related Tickets & Documents

- Related Issue https://github.com/forem/forem/issues/20090
- Closes #

## Added/updated tests?
_We encourage you to keep the code coverage percentage at 80% and above._

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

### CODE REVIEW

File: app/controllers/async_info_controller.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@
 class AsyncInfoController < ApplicationController
   NUMBER_OF_MINUTES_FOR_CACHE_EXPIRY = 15
   before_action :set_cache_control_headers, only: %i[navigation_links]
+  before_action :store_language_accept_header, only: %i[base_data]
```
Comment (maestromac) [inline app/controllers/async_info_controller.rb:5]: I don't think updating DB on a GET request is a good idea.

Comment (maestromac) [inline app/controllers/async_info_controller.rb:5]: I don't think updating DB on a GET request is a good idea.

File: db/migrate/20231003071434_add_browser_accept_language_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class AddBrowserAcceptLanguageToUsers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :users, :browser_accept_language, :string
+    add_column :users, :browser_accept_language_updated_at, :datetime
```
Comment (maestromac) [inline db/migrate/20231003071434_add_browser_accept_language_to_users.rb:4]: While convenient, I'm trying to avoid increasing the size of the users' column, especially when the column may not directly be related to 'user'. I also see possibility of it changing base on user's devices or choice of browsers.

**Review by maestromac — COMMENTED** [review]
I think a more appropriate place to put this is with Ahoy. I am not sure how much of it we're using at the moment, but this is a feature that is readily available on that gem https://github.com/ankane/ahoy#track-additional-data

**Review by lightalloy — COMMENTED** [review]
Agree with @maestromac on using ahoy visits, and using that data later to set user languages.

**Comment by benhalpern** [comment]
@maestromac @lightalloy Great thoughts — definitely Ahoy infrastructure is the place to lean on here.

My only big concern would be that the ahoy table is currently our biggest table at 98GB _while being super lean on a per-row-basis_. So adding a whole string column is going to balloon this.

I definitely think this should tie in to Ahoy, but I'm going to give some extra thought into an approach which might satisfy these needs in the most satisfying way.

**Comment by benhalpern** [comment]
Okay @maestromac @lightalloy I have pushed up the base for a new approach — I'm converting to draft because it's not fully complete.

But the idea is that we create a new table which stores the data for a certain visit context, which are going to generally be unique across many visits, rather than storing a bunch of data on each ahoy visit. It's also only running on _user_ visits, as opposed to anonymous visits, because that is the data most relevant to us here.

Basically, a user might have 20 visits in one context — then they change devices or countries, and they start aggregating on the new context, but we store all of these as a reference vs on each ahoy visit.

Thanks for the feedback, I think this is a much better direction than originally proposed.

**Comment by lightalloy** [comment]
> My only big concern would be that the ahoy table is currently our biggest table at 98GB while being super lean on a per-row-basis. So adding a whole string column is going to balloon this.

Separately from this particular pr/feature. I wonder do we need to store all this data. Maybe we could "combine" the stored data and store some kind of reports that we need separately, and remove the data from the table itself :thinking:

**Comment by benhalpern** [comment]
@lightalloy yeah, definitely. We can roll up and/or delete old data in this table. Just need to commit to a policy.

**Review by maestromac — APPROVED** [review]
This looks great!
